### PR TITLE
Add memcache memoize support.

### PR DESCRIPTION
**Pull request by s-sathish**
Fixes # .
Added support for caching results in Memcache

#### What does this PR do?
Added support for caching results in Memcache

#### Checklist
- [  ] Pull request contains a clear definition of changes
Code comments are added, README.md is updated.
- [  ] Tests (either unit, integration, or acceptance) written and passing
- [  ] Relevant documentation produced and/or updated
README.md is updated with the relevant documentation for memcache support.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -14,7 +14,9 @@
         "sort-packages": true
     },
     "require": {
-        "php": "^7.0"
+        "php": "^7.0",
+        "ext-memcache": "*",
+        "ext-json": "*"
```
Comment (chrisryan) [inline composer.json]: This is not needed as `json_encode` and `json_decode` are part of the core php.

File: composer.json [hunk 2]
```diff
@@ -14,7 +14,9 @@
         "sort-packages": true
     },
     "require": {
-        "php": "^7.0"
+        "php": "^7.0",
+        "ext-memcache": "*",
```
Comment (chrisryan) [inline composer.json]: This should be in a suggest block so people who do not use memcache are not required to install it.

File: composer.json (hunk 1 above)
Comment (chadicus) [reply composer.json]: The json extension is part of core php, however it can be disabled. I suggest keeping the requirement

Comment (s-sathish) [reply composer.json]: Yeah, Removed "ext-json" from the list.

File: composer.json (hunk 2 above)
Comment (s-sathish) [reply composer.json]: Moved it to the require-dev section + added info in suggest block.

**Comment by s-sathish** [comment]
> Tests need to be added and the travis ci file needs to be updated to include memcache in the test environment as currently the builds are all failing because of that composer dependency.

@chrisryan Tests for memcache memoize is added.

File: composer.json (hunk 1 above)
Comment (s-sathish) [reply composer.json]: Added back the ext-json require. Anyways it doesn't make any difference.

**Comment by chadicus** [comment]
@s-sathish  in order for the tests to pass on Travis-CI, you'll need to update the .travis.yml file.  I'm not 100% what should be added, but it should be similar to the following.
```
services:
   - memcached
```
and
```
before_install: echo "extension=memcached.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
```
You can find more information here: https://docs.travis-ci.com/user/database-setup/#memcached

**Comment by s-sathish** [comment]
> @s-sathish in order for the tests to pass on Travis-CI, you'll need to update the .travis.yml file. I'm not 100% what should be added, but it should be similar to the following.
> 
> ```
> services:
>    - memcached
> ```
> 
> and
> 
> ```
> before_install: echo "extension=memcached.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
> ```
> 
> You can find more information here: https://docs.travis-ci.com/user/database-setup/#memcached



> @s-sathish in order for the tests to pass on Travis-CI, you'll need to update the .travis.yml file. I'm not 100% what should be added, but it should be similar to the following.
> 
> ```
> services:
>    - memcached
> ```
> 
> and
> 
> ```
> before_install: echo "extension=memcached.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
> ```
> 
> You can find more information here: https://docs.travis-ci.com/user/database-setup/#memcached

I referred to the below and tried setting up the .travis.yml file. Not sure if that will work correct.
https://blog.travis-ci.com/2013-03-08-preinstalled-php-extensions

File: .travis.yml [hunk 3]
```diff
@@ -7,11 +7,15 @@ php:
...
   - PREFER_LOWEST="--prefer-lowest --prefer-stable"
   - PREFER_LOWEST=""
+services:
+  - memcached
 matrix:
   fast_finish: true
   allow_failures:
     - php: nightly
 before_script:
+  - phpenv config-add ./build/php/memcache.ini
```
Comment (chadicus) [inline .travis.yml]: I don't think you need this line.

**Comment by s-sathish** [comment]
Now we get this - 
`PHP Warning:  PHP Startup: Unable to load dynamic library '/home/travis/.phpenv/versions/7.0.33/lib/php/extensions/no-debug-zts-20151012/memcache.so' - /home/travis/.phpenv/versions/7.0.33/lib/php/extensions/no-debug-zts-20151012/memcache.so: cannot open shared object file: No such file or directory in Unknown on line 0`

@chadicus I'm not sure if adding the following will work or not.
before_script:
  - pecl install memcache

i.e we try to install the php-memcache extension and then proceed with the tests.

**Comment by s-sathish** [comment]
Not sure if it is because the xenial build.
https://travis-ci.community/t/unable-to-load-dynamic-library-memcached-so/2232

**Comment by chrisryan** [comment]
I am not familiar with the syntax for doing this but something is likely missing: Here is the docs I found related to adding extensions: https://docs.travis-ci.com/user/languages/php/#installing-additional-php-extensions

**Comment by s-sathish** [comment]
Yeah, That was also tried.
https://github.com/traderinteractive/memoize-php/pull/22/commits/c914269539127c8eb9b8d0e1fdafdde73883f94a

But the installation of php memcache extension itself fails, preventing further steps.

**Comment by kmcurry** [comment]
@chadicus this PR is 701 days old. Can it be closed?

**Comment by chadicus** [comment]
@kmcurry this is not my pull request

**Pull request closed without merging**